filebeat/input/v2/input-cursor/clean.go

```diff
@@ -0,0 +1,45 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package cursor
+
+import (
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/go-concert/unison"
+)
+
+// cleaner removes finished entries from the registry file.
```

Comment:
If I understand correctly the finished states are not claimed by any input and there is no pending state update. So this means that as long as Filebeat is running it must claim the resource in order to persist its state. Does this mean that Filebeat cannot close an input source temporarily e.g. a journal to wait until new entries show up otherwise its state is lost? Or is this going to implement functionality similar to those of the options `clean_*` for `log` input?

Reply:
> finished states are not claimed by any input and there is no pending state update.

correct.

> So this means that as long as Filebeat is running, it must claim the resource to persist its state.

The resource is claimed by an active input. If there is no active input (e.g. autodiscovery closed the input after the container has been deleted, or the configuration file has changed between restarts), it is 'free' and might be removed by the cleaning process.

> Does this mean that Filebeat cannot close an input source temporarily,

More or less correct. With the implementation of this input manager we add a TTL to each entry in the store. unclaimed resources are removed if `latest update + TTL < now`.

This input manager requires sources to be configured statically and upfront based on the configuration. Internally resources can't be released and reclaimed without an external signal.
But:
- the input.Run method is allowed to close connections, wait, and reopen. The resource is claimed for as long as the go-routine is alive (Run did not return).
- We might consider to add a 'flag' to a resource saying that it can not be cleaned yet. This flag would be active in  memory only and reset on beats restart

Filebeat can have different input manager instances and even different input manager implementations that can all coexist and provide different coordination behavior. For the log input we can not configure `[]Source` upfront, as we need a discovery mechanism (file watcher) adding and removing sources dynamically. Here we have two options, enhance this input manager, or develop a second input manager that can handle dynamic sources. For dynamic sources we should also consider some signaling like 'source removed' or 'source renamed', so we implement `close_removed` and `clean_removed` on top of these signals.

Almost all stateful filebeat inputs we have do configure the source statically at init-time. Even the windows event logs and the journald log (configure system_logs, path, or a filename). In the later case the journald libraries do the heavy lifting for use (based on our configured source).
AFAICT the only special input is the log input.

> Or is this going to implement functionality similar to those of the options clean_* for log input?

Yes. The TTL and cleaner are similar to `clean_inactive`. This input manager does not have a `clean_removed`, as the sources are 'static' and we have no way to tell it that a resource is gone for good. 

For an input manager that supports dynamic sources, having a file watcher we could emit a `remove` immediately, even cancelling pending update operations (or set TTL to 0 and wait for the resource to be 'finished').


For an example usage check out the journald input implementation: https://github.com/urso/beats/blob/fb-input-v2-combined/filebeat/features/input/journald/input.go
The journald input creates it's own InputManager, that coordinates journald input instances only.